book/src/pong-tutorial/pong-tutorial-05.md

```diff
@@ -0,0 +1,887 @@
+            "0".to_string(),
+            [1., 1., 1., 1.],
+            50.,
+        )).build();
+
+    world.add_resource(ScoreText { p1_score, p2_score });
+}
+```
+
+Here, we add some UI imports and create a new `initialise_scoreboard` function,
+which we'll call in the `on_start` method of the `Pong` gamestate.
+
+Inside `initialise_scoreboard`, we're first going to load up a font which we've
+saved to `font/square.ttf` ([download][font-download]) in the app root. We pull
+in the `TtfFormat` to match this font type, load the font as a resource in the
+world, and then save the handle to a `font` variable (which we'll use to create
+our `UiText` components).
+
+Next, we create a transform for each of our two scores by giving them a unique
+id (`P1` and `P2`), a UI `Anchor` at the top middle of our window, and then
+adjust their global `x`, `y`, and `z` coordinates, `width`, `height`, and
+`tab-order`.
+
+After creating the `font` and `transform`s, we'll create an `Entity` in the
+world for each of our players' scores, with their `transform` and a `UiText`
+component (with a `font` handle, initial `text`, `color`, and `font_size`).
+
+Finally, we initialize a `ScoreText` structure containing each of our UI
+`Entity`s and add it as a resource to the world so we can access it from our
+`System`s later.
+
+If we've done everything right so far, we should see `0` `0` at the top of our
+game window. You'll notice that the scores don't update yet when the ball makes
+it to either side, so we'll add that next!
+
+[font-download]: https://github.com/amethyst/amethyst/raw/master/examples/assets/font/square.ttf
+
+
+## Updating the Scoreboard
+
+All that's left for us to do now is update the UI whenever a player scores a
+point. You'll see just how easy this is with our `ECS` design. All we have to do
+is modify our `WinnerSystem` to access the players' scores and update them
+accordingly:
+
+```rust,no_run,noplaypen
+# extern crate amethyst;
+#
+# mod pong {
+#     use amethyst::ecs::prelude::*;
+#
+#     pub struct Ball {
+#         pub radius: f32,
+#         pub velocity: [f32; 2],
+#     }
+#     impl Component for Ball {
+#         type Storage = DenseVecStorage<Self>;
+#     }
+#
+#     #[derive(Default)]
+#     pub struct ScoreBoard {
+#         pub score_left: i32,
+#         pub score_right: i32,
+#     }
+#
+#     pub struct ScoreText {
+#         pub p1_score: Entity,
+#         pub p2_score: Entity,
+#     }
+#
+#     pub const ARENA_WIDTH: f32 = 100.0;
+# }
+#
+use amethyst::{
+#     core::transform::Transform,
+    // --snip--
+    ecs::prelude::{Join, ReadExpect, System, Write, WriteStorage},
+    ui::UiText,
+};
+
+use pong::{Ball, ScoreBoard, ScoreText, ARENA_WIDTH};
+
+pub struct WinnerSystem;
+
+impl<'s> System<'s> for WinnerSystem {
+    type SystemData = (
+        WriteStorage<'s, Ball>,
+        WriteStorage<'s, Transform>,
+        WriteStorage<'s, UiText>,
+        Write<'s, ScoreBoard>,
+        ReadExpect<'s, ScoreText>,
```

Comment:
Maybe explain why Write and ReadExpect are used here? I think it's the first time it come out in the pong tutorial.

Reply:
Sure thing, I'll look those up too, hahah.